File: src/main/java/org/dedda/bratwurst/lang/BWFunction.java
package org.dedda.bratwurst.lang;

import org.dedda.bratwurst.lang.scope.Scope;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Created by dedda on 10/9/15.
 *
 * @author dedda
 */
public class BWFunction extends BWExpression {

    private final String name;
    private final BWInstruction[] instructions;
    private final List<BWVariable> variables = new ArrayList<>();
    private BWObject value = new BWInteger(0);

    public BWFunction(String name, BWInstruction[] instructions) {
        super(0);
        this.name = name;
        this.instructions = instructions;
    }

    @Override
    public BWObject getValue() {
        return value;
    }

    @Override
    public int getIntValue() {
        return value.getIntValue();
    }

    @Override
    public String getValueType() {
        return value.getValueType();
    }

    @Override
    public void run(Scope scope) {
        variables.addAll(Arrays.asList(getArguments()));
        for (BWVariable argument : getArguments()) {
            argument.run(scope);
        }
        for (BWInstruction instruction : instructions) {
            instruction.run(scope);
            if (instruction instanceof Return) {
                value = ((Return) instruction).getValue();
                break;
            }
        }
    }

    public String getName() {
        return name;
    }

    public BWInstruction[] getInstructions() {
        return instructions;
    }

    public List<BWVariable> getVariables() {
        return variables;
    }
}

File: src/main/java/org/dedda/bratwurst/lang/BWFunction.kt
package org.dedda.bratwurst.lang

import org.dedda.bratwurst.lang.scope.Scope

/**
 * Created by dedda on 10/9/15.
 *
 * @author dedda
 */
class BWFunction(val name: String, val instructions: List<BWInstruction>) : BWExpression(0) {
    private val variables: MutableList<BWVariable> = ArrayList()
    private var value: BWObject = BWInteger(0)

    override fun getValue(): BWObject {
        return value
    }

    override fun getIntValue(): Int {
        return value.intValue
    }

    override fun getValueType() = value.valueType

    override fun run(scope: Scope) {
        variables.addAll(listOf(*arguments))
        for (argument in arguments) {
            argument.run(scope)
        }
        for (instruction in instructions) {
            instruction.run(scope)
            if (instruction is Return) {
                value = instruction.value
                break
            }
        }
    }
}